ruby/ql/lib/codeql/ruby/frameworks/core/String.qll

```diff
@@ -0,0 +1,542 @@
+      ) and
+      preservesValue = false
+    }
+  }
+
+  /**
+   * A flow summary for the `String#chomp`, `String#chomp!`, `String#chop` and `String#chop!` methods.
+   */
+  private class ChompSummary extends SummarizedCallable {
+    private MethodCall mc;
+
+    ChompSummary() { this = ["chomp", "chomp!", "chop", "chop!"] and mc.getMethodName() = this }
+
+    override MethodCall getACall() { result = mc }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "Receiver" and
+      preservesValue = false and
+      (
+        output = "ReturnValue"
+        or
+        this = ["chomp!", "chop!"] and output = "Receiver"
+      )
+    }
+  }
+
+  // TODO: we already have a summary for Array#clear. Check that it applies correctly to String#clear.
+  /**
+   * A flow summary for `String#concat` and `String#prepend`.
+   */
+  private class ConcatSummary extends SimpleSummarizedCallable {
+    ConcatSummary() {
+      // `concat` and `prepend` omitted because they clash with the summaries for
+      // `Array#concat` and `Array#prepend`.
+      // this = ["concat", "prepend"]
+      none()
+    }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = ["Receiver", "Argument[_]"] and
+      output = ["ReturnValue", "Receiver"] and
+      preservesValue = true
+    }
+  }
+
+  /**
+   * A flow summary for `String#delete(!)`, `String#delete_prefix(!)` and `String#delete_suffix(!)`.
+   */
+  private class DeleteSummary extends SimpleSummarizedCallable {
+    DeleteSummary() { this = ["delete", "delete_prefix", "delete_suffix"] + ["", "!"] }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      taintIdentityFlow(input, output, preservesValue)
+    }
+  }
+
+  /**
+   * A flow summary for `String#downcase(!)`, `String#upcase` and `String#swapcase(!)`.
+   */
+  private class DowncaseSummary extends SimpleSummarizedCallable {
+    DowncaseSummary() { this = ["downcase", "upcase", "swapcase"] + ["", "!"] }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      taintIdentityFlow(input, output, preservesValue)
+    }
+  }
+
+  /**
+   * A flow summary for `String#dump` and `String#undump`.
+   */
+  private class DumpSummary extends SimpleSummarizedCallable {
+    DumpSummary() { this = ["dump", "undump"] }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      taintIdentityFlow(input, output, preservesValue)
+    }
+  }
+
+  /**
+   * A flow summary for `String#each_line` and `String#lines`.
+   */
+  private class EachLineSummary extends SummarizedCallable {
+    MethodCall mc;
+
+    EachLineSummary() { this = ["each_line", "lines"] and mc.getMethodName() = this }
+
+    override MethodCall getACall() { result = mc }
+
+    predicate hasBlock() { exists(Block b | b = mc.getBlock()) }
```

Comment:
There's a common misunderstanding about how fields work in CodeQL, which ultimately is why this predicate won't work as expected.

Declaring a field in the class is basically syntactic sugar for wrapping every single predicate body with an extra `exists` clause with some conditions attached to it. It doesn't mean the value of `this` is associated with unique field values.

The `hasBlock()` predicate holds if _any_ call to `each_line` or `lines` has a block argument. Doing `this.hasBlock()` in a member predicate does in any way guarantee that the value of `mc`, as seen by that predicate, actually has a block argument.

The only way to handle this is to make sure the value of `this` encodes knowledge of whether the block argument exists. Then `getACall()` should only find the calls where the block argument respectively exists or doesn't exist.

If you want to use a field, it could look like this:
```codeql
  private class EachLineSummary extends SummarizedCallable {
    string hasBlock;

    EachLineSummary() { this = "each_line_" + hasBlock and hasBlock = ["block", "noblock"] }

    override MethodCall getACall() {
      result.getMethodName() = ["each_line", "lines"] and
      if exists(result.getBlock()) then
        hasBlock = "block"
      else
        hasBlock = "noblock"
    }

    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
      preservesValue = false and
      input = "Receiver" and
      (
        output = "Parameter[0] of BlockArgument"
        or
        hasBlock = "noblock" and output = "ArrayElement[?] of ReturnValue"
        or
        hasBlock = "block" and output = "ReturnValue"
      )
    }
```

It may also be nicer to split this into two classes, but importantly, the two classes still need to have distinct `this` values.

Note that the `getACall()` and `propagatesFlowExt` predicate are separate because, by design, the `propagatesFlowExt` should not depend on the precise call-site. This is because `propagatesFlowExt` must be computed an an earlier stage than `getACall()`. (`getACall()` can in general depend on the data-flow IR, and the data-flow IR depends on `propagatesFlowExt`).

Reply:
Ugh, yeah. This is not the first time I've made this mistake and probably won't be the last. Splitting into multiple classes sounds like the way to go.